### PR TITLE
feat(elbv2): support `UDP` and `TCP_UDP` protocols

### LINKED ISSUE #3107: Add support for UDP protocol in NLB
The NLB now supports the UDP protocol: https://aws.amazon.com/about-aws/whats-new/2019/06/network-load-balancer-now-supports-udp-protocol/.  Please add support for it in the CDK.

### LINKED ISSUE #4341: Support for TCP_UDP as a supported protocol
It's currently possible to create a network load balancer listener that supports dual protocols of TCP and UDP by setting the protocol to TCP_UDP. 

"TCP_UDP" protocol is supported by CloudFormation for [AWS::ElasticLoadBalancingV2::Listener](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticloadbalancingv2-listener.html#cfn-elasticloadbalancingv2-listener-protocol)

Currently, the `enum` for Protocols only allows HTTP, HTTPS, TCP, or TLS, but doesn't support the value 'UDP' or 'TCP_UDP' - [code path](https://github.com/aws/aws-cdk/blob/19ae072be9788a0c1ae7d3a347a99b3f3494f010/packages/%40aws-cdk/aws-elasticloadbalancingv2/lib/shared/enums.ts#L19)

Workaround: Using Escape Hatches - https://docs.aws.amazon.com/en_pv/cdk/latest/guide/cfn_layer.html#cfn_layer_raw

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by nmussy**
Adds support for `UDP` and `TCP_UDP` protocol for NLB and health checks
* Add `UDP` and `TCP_UDP` to `Protocol`
* Document `Protocol` to indicate that `HTTP` and `HTTPS` cannot be used for NLB
* Validate ALB health check protocols
* Fix NLB health check protocol validation (was `TLS|TCP`, now `HTTP|HTTPS|TCP`)
* Fix NLB test cases

References:
* [NLB health checks](https://docs.aws.amazon.com/elasticloadbalancing/latest/network/target-group-health-checks.html)
* [ALB health checks](https://docs.aws.amazon.com/elasticloadbalancing/latest/application/target-group-health-checks.html)

Fixes #4341
Fixes #3107

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by nmussy** [comment]
I preferred creating a new `NetworkProtocol` enum, to align it with the existing `ApplicationProtocol`. Tell me if you'd prefer we kept the existing `Protocol` for NLB.

Another consistency change we could do would be to rename `Protocol` into `HealthCheckProtocol`, but I'm not sure if it's worth another breaking change.

File: packages/@aws-cdk/aws-elasticloadbalancingv2/lib/nlb/network-listener.ts [hunk 1]
```diff
@@ -26,7 +26,7 @@ export interface BaseNetworkListenerProps {
    *
    * @default - TLS if certificates are provided. TCP otherwise.
    */
-  readonly protocol?: Protocol;
+  readonly protocol?: NetworkProtocol;
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-elasticloadbalancingv2/lib/nlb/network-listener.ts]: Sorry, but we can't have this. I don't think this is a good enough reason to break API compatibility.

I would be more in favor of adding `TCP_UDP` to the existing `Protocol` enum, noting in the docstring that it is only for NLBs, and adding runtime validation to make sure it isn't being used for ALBs.

**Comment by nmussy** [comment]
Just noticed I messed up something, I'll fix it and reopen, sorry

File: packages/@aws-cdk/aws-elasticloadbalancingv2/lib/nlb/network-target-group.ts [hunk 2]
```diff
@@ -106,8 +106,8 @@ export class NetworkTargetGroup extends TargetGroupBase implements INetworkTarge
     if (healthCheck.path) {
       ret.push('Health check paths are not supported for Network Load Balancer health checks');
     }
-    if (healthCheck.protocol && healthCheck.protocol !== Protocol.TCP && healthCheck.protocol !== Protocol.TLS) {
-      ret.push(`Health check protocol '${healthCheck.protocol}' is not supported. Must be one of [TCP, TLS]`);
```
Comment (nmussy) [inline packages/@aws-cdk/aws-elasticloadbalancingv2/lib/nlb/network-target-group.ts]: This isn't the case according to the docs [the docs](https://docs.aws.amazon.com/elasticloadbalancing/latest/network/target-group-health-checks.html):

> HealthCheckProtocol
> The possible protocols are HTTP, HTTPS, and TCP

**Pull request merged**